text/0095-cognito-construct-library/working-backwards-readme.md

```diff
@@ -0,0 +1,567 @@
+
+User pools can be configured to enable MFA. It can either be turned off, set to optional or made required. Setting MFA
+to optional means that individual users can choose to enable it. Phone numbers must be verified if MFA is enabled.
+Additionally, MFA can be sent either via SMS text message or via a time-based software token.
+See the [documentation on MFA](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-mfa.html) to
+learn more.
+
+This can be configured by setting the `mfa.enforcement` option under `security` properties to be one of the values in
+the `MfaEnforcement` enum. Available values are `REQUIRED`, `OPTIONAL`, `OFF`.
+The type of MFA can be configured by its peer property `type` which can be set to a list of values in the enum
+`MfaType`. The available values are `SMS` and `SOFTWARE_TOKEN`.
+
+User pools can be specify the constraints that should be applied when users choose their password. Constraints such as
+minimum length, whether lowercase, numbers and/or symbols are required can be specified.
+
+In order to send an SMS, Cognito needs an IAM role that it can assume with permissions that allow it to send an SMS on
+behalf of the AWS account. By default, CDK will create this IAM user but allows for it to be overridden via the
+`smsRole` permissions.
+
+> These are the defaults that will be part of tsdoc, and not part of the README -
+> * security.mfa.enforcement: OPTIONAL
+> * security.mfa.type: SMS
+> * security.passwordPolicy.minLength: 8
+> * security.passwordPolicy.required - lowercase, numbers and symbols
+> * security.smsRole - assumable by `cognito-idp.amazonaws.com` and permitting `sns:Publish` action against resource `*`.
+>   The role assumption will be conditioned on a strict equals on an ExternalId that will be unique to this user pool.
+
+Code sample:
+
+```ts
+new UserPool(this, 'myuserpool', {
+  // ...
+  // ...
+  security: {
+    mfa: {
+      enforcement: MfaEnforcement.REQUIRED,
+      type: [ MfaType.SMS, MfaType.SOFTWARE_TOKEN ]
+    },
+    passwordPolicy: {
+      required: [ PasswordPolicy.LOWERCASE, PasswordPolicy.NUMBERS, PasswordPolicy.SYMBOLS ],
+      minLength: 12,
+    },
+    smsRole: iam.Role.fromRoleArn(/* ... */),
+  }
+});
+```
+
+> Internal Note: Password policy via UserPool-Policies, MFA enable via UserPool-MfaConfiguration; MFA type via
+> UserPool-EnabledMfas; smsRole via UserPool-SmsConfiguration
+
+> Internal Note: Account Recovery settings are missing from UserPool CloudFormation resource.
+
+### Emails
+
+Cognito will handle sending emails to users in the user pool. The address from which emails are sent can be configured
+on the user pool via the `from` property, and the `replyTo` property to configure the email where replies are sent.
+
+User pools can also be configured to send emails through Amazon SES, however, that is not yet supported via the CDK. Use
+the cfn layer to configure this - https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html.
```

Comment:
Why isn't this going to be supported? Seems like it could be a simple switch.

Reply:
Yes if CFN can do it there shouldn't really be a blocker.

Or is it because SES doesn't have appropriate L2s yet?

Reply:
> Or is it because SES doesn't have appropriate L2s yet?

Correct! 

> Seems like it could be a simple switch.

It's not a simple switch. A bunch of configurability [comes along with it](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cognito-userpool-emailconfiguration.html#aws-properties-cognito-userpool-emailconfiguration-properties) and without proper class modeling in SES, this will look weird.

At the same time, given the extended nature of Cognito's features, something had to be cut for the first round of implementation and this made the cut. This is not saying we will not support it but just not yet. Users can always avail this configuration via escape hatches, albeit it won't be clean.